core/invoker/src/main/scala/whisk/core/containerpool/ContainerProxy.scala

```diff
@@ -248,16 +250,20 @@ class ContainerProxy(
           // Sending the message to self on a failure will cause the message
           // to ultimately be sent back to the parent (which will retry it)
           // when container removal is done.
-          case Failure(_) => self ! job
+          case Failure(_) =>
+            rescheduleJob = true
+            self ! job
         }
         .flatMap(_ => initializeAndRun(data.container, job))
         .map(_ => WarmedData(data.container, job.msg.user.namespace, job.action, Instant.now))
         .pipeTo(self)
 
       goto(Running)
 
-    // timeout or removing
-    case Event(StateTimeout | Remove, data: WarmedData) => destroyContainer(data.container)
+    // container is reclaimed by the pool or it has become too old
+    case Event(StateTimeout | Remove, data: WarmedData) =>
+      rescheduleJob = true // to supress sending message to the pool and not double count
+      destroyContainer(data.container)
```

Comment:
I don't think it's guaranteed that the job is in fact rescheduled in this case, which would lead to a nasty lock situation if too many containers are reclaimed due to timeouts.

In fact I think this specific situation needs to be resolved differently, since at the point of this transition you won't know whether or not you'll resend a job?

Reply:
If a container is reclaimed because of a timeout only, it does not reflect a change in the pool the capacity. It is sufficient to instruct the parent to remove the container from it’s free pool.

What do you mean by it’s not guaranteed that the job is rescheduled in case of a timeout (note there is a test specifically for this case)?

Reply:
> Ideally you'd want Line 272 to set the flag and then eventually send the right message. The message is sent before Line 272 might ever be used though.

I had this thought too, to send a `RescheduledRun` message instead of the current `Run` but the state machine transitions would require a more invasive change, I think, and I abandoned it. Want to try it?

Reply:
Ah you're right indeed. The "timeout means the capacity is already freed earlier" is the trick here.

I might've expressed myself badly: Ideally you'd want Line 272, but it's impossible to do there.